warehouse/macaroons/caveats.py

```diff
@@ -145,9 +146,30 @@ def verify(self, key):
         self.verifier.satisfy_general(ProjectIDsCaveat(self))
 
         try:
-            return self.verifier.verify(self.macaroon, key)
+            result = self.verifier.verify(self.macaroon, key)
         except (
             pymacaroons.exceptions.MacaroonInvalidSignatureException,
             Exception,  # https://github.com/ecordell/pymacaroons/issues/50
-        ):
-            return False
+        ) as exc:
+            failure_reasons = []
+            for cb in self.verifier.callbacks:
+                failure_reason = getattr(cb, "failure_reason", None)
+                if failure_reason is not None:
+                    failure_reasons.append(failure_reason)
+
+            # If we have a more detailed set of failure reasons, use them.
+            # Otherwise, use whatever the exception gives us.
+            if len(failure_reasons) > 0:
+                return WarehouseDenied(
+                    ", ".join(failure_reasons), reason="invalid_api_token"
+                )
+            else:
+                return WarehouseDenied(str(exc), reason="invalid_api_token")
```

Comment:
I'm not sure that we want to just expose ``str(exc)`` to end users like this? It's *probably* fine, but who knows what's crammed in some random exception string?

Reply:
Good point -- at the very least it won't be a helpful message when it's just an `Exception` and not a `MacaroonInvalidSignatureException`. I can constrain it to the latter exception and use a generic message for the `Exception` case, if that sounds good to you?

Reply:
[6826cf9](https://github.com/pypi/warehouse/pull/11885/commits/6826cf9aac86d8a2d863cc8a070820b7d146bbef) eliminates this, except for when we're stringifying the signature exception.